cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java

```diff
@@ -150,11 +159,20 @@ Set<ProgramRunId> terminatePrograms() {
         // If this run was initiated by this CDAP instance to run on another instance, it will use the tethering
         // provisioner. The instance that actually runs it on the other side will have a non-null peer name.
         String provisionerName = SystemArguments.getProfileProvisioner(record.getSystemArgs());
+        ProgramRunId programRunId = record.getProgramRunId();
         if (record.getPeerName() == null && TetheringProvisioner.TETHERING_NAME.equals(provisionerName)) {
+          if (record.getStoppingTs() != null &&
+            currentTimeInSecs - tetheredProgramTerminateTimeSecs > record.getStoppingTs() &&
+            record.getTerminateTs() != null && record.getTerminateTs() <= currentTimeInSecs) {
```

Comment:
this check already happens in the Predicate passed to store.getRuns().

It would be better if we move this entire check into the createFilter() method, that way all the logic around what to force terminate is in the same place. Plus we use less memory by filtering on the fly.

Reply:
done